### DISCUSSION TITLE
Component not ssr loaded

### DISCUSSION THREAD

**Discussion by abhikumar22** [vercel/next.js#27699] [Help]
I have one page.
In that i have one component which is 
dynamically loaded.

In that component in useeffect i called an API and rendered the data.

But that page is not SSR even though i used dynamic loading and also set ssr to false..

Any help here ?

**Comment by mrr11k** [comment]
Hi, you should use [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering) (or even [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation)) to fetch data on the server or (even at build time).

useEffect is always executed on the client side.

**Reply by abhikumar22** [reply]
Or if o use componentdidmount or something else?

**Reply by mrr11k** [reply]
componentDidMount is a life-cycle method of a class based React component. useEffect can do the same for functional React components. Check out [this](https://reactjs.org/docs/hooks-effect.html) for a good explanation.

But both componentDidMount for class based components and useEffect for functional components are executed on the client side. 

This is why you have to use getServerSideProps to fetch data server-side.

**Reply by abhikumar22** [reply]
Lets say we have to seperate component with their api's, then we will not be able to do this. It will be client rendered always.

To avoid this we will have to call each and every api on getServerSideProps of that page and pass data as props to the components then only it will be server rendered ?
@mrr11k

**Reply by mrr11k** [reply]
Yes, there is no way around getServerSideProps to have the data ready on the server.

One other option could be to render e.g. [skeletons](https://material-ui.com/components/skeleton/) on the server and then replace them client side.

**Reply by abhikumar22** [reply]
@mrr11k 
loading is not the problem, i need SEO friendly pages, that's why :)

**Reply by mrr11k** [reply]
You might also checkout [Incremental Static Regeneration](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration).

**Reply by abhikumar22** [reply]
Thanks for the help :)
